dags/utils/gcp.py

```diff
@@ -265,3 +265,83 @@ def reprocess_parquet(parent_dag_name,
             DummyOperator(task_id='no_reprocess')
 
         return dag
+
+
+def export_to_parquet(
+    table,
+    arguments=[],
+    dag_name="export_to_parquet",
+    parent_dag_name=None,
+    default_args=None,
+    gcp_conn_id="google_cloud_derived_datasets",
+    dataproc_zone="us-central1-a",
+    dataproc_storage_bucket="moz-fx-data-derived-datasets-parquet",
+    num_workers=1,
+    num_preemptible_workers=0,
+):
+
+    """ Export a BigQuery table to Parquet.
+
+    https://github.com/mozilla/bigquery-etl/blob/master/script/pyspark/export_to_parquet.py
+
+    :param str table:                             [Required] BigQuery table name
+    :param List[str] arguments:                   Additional pyspark arguments
+    :param str dag_name:                          Name of DAG
+    :param Optional[str] parent_dag_name:         Parent DAG name
+    :param Optional[Dict[str, Any]] default_args: DAG configuration
+    :param str gcp_conn_id:                       Airflow connection id for GCP access
+    :param str dataproc_storage_bucket:           Dataproc staging GCS bucket
+    :param str dataproc_zone:                     GCP zone to launch dataproc clusters
+    :param int dataproc_num_workers:              Number of Dataproc workers
+    :param int dataproc_num_preemptible_workers:  Number of Dataproc preemptible workers
+
+    :return: airflow.models.DAG
+    """
+
+    cluster_name = table.replace("_", "-") + "-{{ ds_nodash }}"
+    dag_prefix = parent_dag_name + "." if parent_dag_name else ""
+    connection = GoogleCloudBaseHook(gcp_conn_id=gcp_conn_id)
+
+    with models.DAG(dag_id=dag_prefix + dag_name, default_args=default_args) as dag:
+
+        create_dataproc_cluster = DataprocClusterCreateOperator(
+            task_id="create_dataproc_cluster",
+            cluster_name=cluster_name,
+            gcp_conn_id=gcp_conn_id,
+            project_id=connection.project_id,
+            properties={
+                "core:fs.s3a.access.key": models.Variable.get("s3_export_access_key"),
+                "core:fs.s3a.secret.key": models.Variable.get("s3_export_secret_key"),
```

Comment:
These variables would require configuring in airflow.

I'm not sure if this is the best way to do this, because this method makes the credentials visible to anyone that can describe the dataproc cluster, and IDK if `models.Variable` is an acceptable place to store even very restricted credentials.

I could alternatively write to a `gs://` location in from spark, and then call a kubernetes operator to copy from GCS to S3, but for clients_last_seen this is 375 GiB being copied, and I would prefer to take advantage of spark's parallelism to do it.

cc @haroldwoo and @jasonthomas

Reply:
Why do we need to write to s3? Having the aws credentials visible to dataproc users may not be that bad if the keys are restricted to only write to a particular bucket. Another option may be to put the creds in a gcs bucket and have the run_dataproc_pyspark task use the file.

Reply:
We are writing back to S3 so that this data is available in our existing AWS tool chains. 

I think doing it this way is fine since the project this will be running in will have limited access. Can we use Airflow Connections here instead to store the AWS credentials [1]? 

When we do provision credentials for this we should limit based on dataset prefix.

[1] https://airflow.readthedocs.io/en/stable/howto/manage-connections.html

Reply:
we need to get these tables to s3 so they are accessible in athena, presto, and spark. This is blocking us from disabling client_count_daily, desktop_dau and desktop_active_dau parquet tables, as well as blocking some MAU-type analysis for an intern project.

as for limiting creds to write a specific bucket, makes sense, though I would probably want to further restrict it to specific paths within the telemetry-parquet bucket, just so that it can't mess with any unexpected tables. it's probably fine because everyone that can see dataproc in that project probably also has full access to the dev iam on aws.

none of this addresses the question of using `models.Variable` to store credentials though. also, would it work to just use `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` from the environment, now that this runs in k8s?

Reply:
those keys are already in use and have a lot more access (full emr, etc.)

Reply:
using a connection seems great, I'll see if I can figure out how to pass connection creds to dataproc

Reply:
okay, pretty sure I'm doing that correctly now.